### DISCUSSION TITLE
Fiber & thread compatibility

### DISCUSSION THREAD

**Discussion by benjaminwood** [ioquatix/ioquatix#28] [Q&A] [answered]
Hi Sam,

First off - thank you for your investment in the Ruby community! I have a general question about fiber & thread compatibility. I bumped into an issue where I was running some code in an async loop, and that code also did some work with threads. Unfortunately, the thread became blocked, and I've had difficulty debugging it. I've researched and read some folks saying that running threads inside a fiber is not a good idea. But I don't feel I have found an authoritative source - could you help settle this matter? 😄 

More generally, I'm writing a gem that attempts to run unknown application code in an async loop. At this point, I'm beginning to wonder if that is a bad idea, given I do not know what the behavior of the code inside the async loop will be. Could you offer some advice for this scenario?

**Comment by benjaminwood** [comment]
Quick follow-up: any blocking IO in the thread will prevent `foo.join(0)` from returning.

```ruby
pipe_name = "/tmp/my_pipe"
system("mkfifo #{pipe_name}")

Async do |task|
  foo = Thread.new do
    puts "foo"
    File.open(pipe_name, 'r') do |pipe|
      # Read the message from the pipe
      message = pipe.gets.chomp
      puts "Received: #{message}"
    end
  end
  puts "bar"
  foo.join(0)
  puts "baz"
end
```

The above code will show bar & foo, but never get to baz. My understanding is that passing 0 to join should cause it to immediately return nil if the thread is not finished. 

Thanks for entertaining these questions. I get the feeling there is something fundamental I do not understand here, and it may have little to do with Async as a library. Appreciate your time!

**Reply by ioquatix** [reply]
It does appear there is some kind of issue with the implementation. I checked the code and it does not work as expected, let me check what's going on.

**Reply by benjaminwood** [reply]
Cool, thanks! I really appreciate it.

**Reply by ioquatix** [reply, marked as answer]
Thanks for the report, I have created the issue: <https://bugs.ruby-lang.org/issues/19709> and have a fix for it locally. I'll try to get this back-ported to Ruby 3.0+.

**Reply by benjaminwood** [reply]
Thank you! I can’t believe you’ve fixed it already 🤯

**Reply by ioquatix** [reply]
Unfortunately back porting this could take 2-3 months. However, it's now in the pipeline. As a short term fix, you could do this:

```ruby
begin
  task.with_timeout(1) do
    thread.join
  end
rescue Async::TimeoutError
  # join timed out
end
```
